### PR TITLE
fix: Jetson/Tegra CUDA optimization and unified memory GPU support

### PR DESCRIPTION
**Pull request by rgthelen**
### **User description**
This PR fixes a 3x performance regression on NVIDIA Jetson devices and adds comprehensive support for unified memory GPU platforms.

## Performance Fixes

- **Root cause fix**: `thinking_budget=0` was being passed instead of `None`, causing ThinkingBudgetProcessor to iterate through 151,936 vocab tokens per generated token even when thinking was disabled

- **Numpy optimization**: Replace Python list iteration with numpy arrays for logits processing, fixing 3x slowdown when thinking mode is enabled

## Results on Jetson Orin Nano 8GB

| Mode | Before | After |
|------|--------|-------|
| Normal | 12.3 tok/s | 36.0 tok/s |
| Thinking | 12.1 tok/s | 34.5 tok/s |

## Code Changes

- Generalize `_is_tegra()` to `_is_unified_memory_gpu()` with:
  - Environment variable override: `LLAMAFARM_SYNC_INFERENCE=1/0`
  - Apple Silicon detection (also unified memory)
  - Better documentation explaining rationale

- Improve inline documentation for:
  - Main-thread backend initialization (crash prevention)
  - Library loading order with RTLD_GLOBAL
  - Memory mapping defaults for unified memory

## Documentation

- Add comprehensive Jetson deployment guide
- Update llamafarm-llama README with Jetson section
- Add platform-specific guide links to deployment docs

___

### **PR Type**
Bug fix, Enhancement

___

### **Description**
- Fix 3x performance regression on Jetson by correcting thinking_budget parameter passing and optimizing logits processing with numpy arrays

- Generalize Tegra detection to unified memory GPU support with environment variable override and Apple Silicon detection

- Optimize library loading order with RTLD_GLOBAL for proper backend registration on Linux/Jetson systems

- Add pre-allocated buffers for detokenization to avoid CFFI thread-safety issues during garbage collection

- Implement synchronous inference mode for unified memory platforms to avoid ThreadPoolExecutor overhead

- Add comprehensive Jetson deployment guide with build instructions, optimization tips, and troubleshooting

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Performance Issues<br/>on Jetson"] --> B["Fix thinking_budget<br/>parameter passing"]
  A --> C["Optimize logits<br/>with numpy arrays"]
  A --> D["Synchronous inference<br/>for unified memory"]
  B --> E["3x speedup<br/>35+ tok/s"]
  C --> E
  D --> E
  F["Library Loading<br/>Issues"] --> G["RTLD_GLOBAL<br/>preload order"]
  G --> H["Proper backend<br/>registration"]
  I["Thread-safety<br/>Issues"] --> J["Pre-allocated<br/>CFFI buffers"]
  J --> K["Stable detokenization"]
  L["Platform Detection"] --> M["Unified memory GPU<br/>detection"]
  M --> N["Jetson/Apple Silicon<br/>support"]
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>_bindings.py</strong><dd><code>Update bindings and improve library loading documentation</code></dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/702/files#diff-a3453a01b75e0cd7c21d1a5d34cd4b22cb627224a8ab5318a570eb3444a4a7d9">+65/-31</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>gguf_language_model.py</strong><dd><code>Add unified memory GPU detection and synchronous inference</code></dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/702/files#diff-aa3d47952f93f3f6d338873bbc86b4f9eb2700b79e17fd3e73e5a2016fc37064">+97/-8</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Bug fix</strong></td><td><details><summary>4 files</summary><table>
<tr>
  <td><strong>llama.py</strong><dd><code>Add pre-allocated buffers and optimize logits processing</code>&nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/702/files#diff-cb2f5fd3e3544c1abb28deac6e74d566628635b5dd4d9bdb64b84f45f2d70559">+50/-28</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>service.py</strong><dd><code>Fix thinking_budget parameter passing to avoid zero value</code></dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/702/files#diff-0d5bb320d0590e4a8fb439fd35b07be371bae7d283b4862979f25fbaad05be14">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>server.py</strong><dd><code>Initialize llama backend in main thread for Jetson stability</code></dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/702/files#diff-eb7e358ca4b36ecd277661e724c825dfbf087b8bbb6aa4accf1a9eb72cc1590f">+40/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>gguf_metadata_cache.py</strong><dd><code>Disable garbage collection during GGUF parsing on aarch64</code></dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/702/files#diff-31e3447bb0c1033d52a68126b973b7f793d6dd4d63c93a2e6732973a93b8f375">+10/-1</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td><strong>index.md</strong><dd><code>Add link to Jetson deployment guide</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/702/files#diff-e82fc477cb532314e446f53034cd3a06449de8b9eb4ab971aef47db214e84301">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>jetson.md</strong><dd><code>Add comprehensive Jetson deployment and optimization guide</code></dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/702/files#diff-23ec800218f610ddf18f361ad88dd28251de9346e34ce0db06bb733efc9100e4">+223/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>README.md</strong><dd><code>Add Jetson build instructions and recommended settings</code>&nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/702/files#diff-2f35030cbdd261c020c23abb49139eac78242badc70be22fb8165955991dd505">+53/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tbody></table>

</details>

___

### CODE REVIEW

**Comment by rgthelen** [comment]
@claude what do you think?

File: docs/website/docs/deployment/jetson.md [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+  -H 'Content-Type: application/json' \
+  -d '{
+    "model": "unsloth/Qwen3-1.7B-GGUF:Q4_K_M",
+    "messages": [{"role": "user", "content": "Hello!"}],
+    "max_tokens": 100,
+    "n_ctx": 2048
+  }'
+```
+
+**Important**: Always pass `n_ctx` in requests to avoid defaulting to the model's full context (40960 for Qwen3), which causes OOM.
```
Comment (mhamann) [inline docs/website/docs/deployment/jetson.md:189]: If someone knows that they're deploying on Jetson, they should just set a sensible value in their llamafarm.yaml project file.

File: packages/llamafarm-llama/src/llamafarm_llama/_bindings.py [hunk 2]
```diff
@@ -134,7 +134,9 @@
         bool no_host;
     };
 
-    // Context parameters - must match llama.cpp b7376+ layout exactly
+    // Context parameters - must match llama.cpp b7690+ layout exactly
+    // Note: Jetson builds use newer llama.cpp (b7690+) which requires the samplers field.
+    // The bundled b7376 binaries for other platforms ignore this extra field safely.
```
Comment (mhamann) [inline packages/llamafarm-llama/src/llamafarm_llama/_bindings.py]: We should just upgrade all platforms to b7690 vs. piecemealing it

File: runtimes/universal/server.py [hunk 3]
```diff
@@ -74,6 +74,46 @@
 logger = UniversalRuntimeLogger("universal-runtime")
 
 
+def _init_llama_backend():
+    """Initialize llama.cpp backend in the main thread.
+
+    CRITICAL FOR STABILITY: On NVIDIA Jetson/Tegra devices with unified memory,
```
Comment (mhamann) [inline runtimes/universal/server.py:108]: More of a side-note, but I think this will be necessary for pytorch and transformers as well on every platform. I occasionally see warnings about something along these lines.

File: packages/llamafarm-llama/src/llamafarm_llama/_bindings.py (hunk 2 above)
Comment (rgthelen) [reply packages/llamafarm-llama/src/llamafarm_llama/_bindings.py]: Done
